packages/@aws-cdk/core/lib/custom-resource-provider.ts

```diff
@@ -0,0 +1,166 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import { AssetStaging } from './asset-staging';
+import { FileAssetPackaging } from './assets';
+import { CfnResource } from './cfn-resource';
+import { Construct } from './construct-compat';
+import { Duration } from './duration';
+import { Size } from './size';
+import { Stack } from './stack';
+import { Token } from './token';
+
+const ENTRYPOINT_FILENAME = '__entrypoint__';
+const ENTRYPOINT_HANDLER = path.join(__dirname, 'custom-resource-provider-handler.js');
+
+/**
+ * Initialization properties for `CustomResourceProvider`.
+ *
+ * @experimental
+ */
+export interface CustomResourceProviderProps {
+  /**
+   * A local file system directory with the provider's code. The code will be
+   * bundled into a zip asset and wired to the provider's AWS Lambda function.
+   */
+  readonly codeDirectory: string;
+
+  /**
+   * The AWS Lambda runtime and version to use for the provider.
+   */
+  readonly runtime: CustomResourceProviderRuntime;
+
+  /**
+   * A set of IAM policy statements to include in the inline policy of the
+   * provider's lambda function.
+   *
+   * @default - no additional inline policy
+   */
+  readonly policyStatements?: any[];
+
+  /**
+   * AWS Lambda timeout for the provider.
+   *
+   * @default Duration.minutes(15)
+   */
+  readonly timeout?: Duration;
+
+  /**
+   * The amount of memory that your function has access to. Increasing the
+   * function's memory also increases its CPU allocation.
+   *
+   * @default Size.mebibytes(128)
+   */
+  readonly memorySize?: Size;
+}
+
+/**
+ * The lambda runtime to use for the resource provider. This also indicates
+ * which language is used for the handler.
+ * @experimental
+ */
+export enum CustomResourceProviderRuntime {
+  /**
+   * Node.js 12.x
+   */
+  NODEJS_12 = 'nodejs12'
+}
+
+/**
+ * An AWS-Lambda backed custom resource provider.
+ *
+ * @experimental
+ */
+export class CustomResourceProvider extends Construct {
+  /**
+   * Returns a stack-level singleton ARN (service token) for the custom resource
+   * provider.
+   *
+   * @param stack The stack in which this provider
+   * @param props Provider properties
+   * @returns the ARN (service token) of the custom resource provider
+   */
+  public static getOrCreate(stack: Stack, resourceType: string, props: CustomResourceProviderProps) {
+    const id = `${resourceType}CustomResourceProvider`;
+    const provider = stack.node.tryFindChild(id) as CustomResourceProvider
+      ?? new CustomResourceProvider(stack, id, props);
+
+    return provider.serviceToken;
+  }
+
+  /**
+   * The service token of this custom resource provider.
+   *
+   * This is the ARN of the provider's AWS Lambda function.
+   */
+  public readonly serviceToken: string;
+
+  public constructor(scope: Construct, id: string, props: CustomResourceProviderProps) {
```

Comment:
Any value in having two entry points? Keep `getOrCreate` as the only entry point.

Reply:
I'll make the ctor private

Reply:
Fixed.